External/FEXCore/Source/Interface/Context/Context.h

```diff
@@ -269,9 +269,23 @@ namespace FEXCore::Context {
 
     uint8_t GetGPRSize() const { return Config.Is64BitMode ? 8 : 4; }
 
+    std::shared_mutex MemoryEntryMutex{};
+    struct MemoryEntry {
+      uint64_t Length;
+      bool Writable;
+    };
+
     void AddNamedRegion(uintptr_t Base, uintptr_t Size, uintptr_t Offset, const std::string &filename);
     void RemoveNamedRegion(uintptr_t Base, uintptr_t Size);
 
+    // Tracks Base Address of maps, Page aligned
+    std::map<uint64_t, MemoryEntry> MemoryMaps;
+    void SetMemoryMap(uintptr_t Base, uintptr_t Size, bool Writable);
+    void ClearMemoryMap(uintptr_t Base, uintptr_t Size);
+
+    void LockBeforeFork();
```

Comment:
They are called before / after fork from the sycall code. I was just replicating the code that is there for some other locks, to avoid some lockups I noticed.

Reply:
(If you don't own a fork before lock, and someone else gets it, then the forked process never unlocks it and will get stuck on next lock attempt)

Reply:
> If I'm writing new code that uses fork(), do I need to use these?
Probably yes, though it depends.

Woops, yes, weird slip there. Better wording:
`If you don't own a lock before fork, and some other thread gets it, then the forked process never unlocks it and will get stuck on next lock attempt`

> What do you mean with "someone else gets it"
Let's say that right before fork, some other thread calls `SetMemoryMap/ClearMemoryMap` and gets the lock, then the fork happens. The new process will end up with the mutex locked, and no way to unlock it.

In https://github.com/FEX-Emu/FEX/blob/main/Source/Tests/LinuxSyscalls/Syscalls/Thread.cpp#L174 there is a workaround for the FD lock, and as I run into the same problem I added this one.

In general, just like with signal masking, this is a problem that can affect any lock, including locks in 3rd party libraries (jemalloc).

For the most part guest applications do fork+execve in multithreading environments, so this is less of a problem in practice.

I think we need to discuss some way to more consistently and systematically handle both of these cases (signals + fork)

Reply:
>  If you don't own a lock before fork, and some other thread gets it, then the forked process never unlocks it and will get stuck on next lock attempt

Ah, interesting. It wasn't obvious to me that fork() only forks the running thread.

I would suggest the following docstring (which only makes sense if we encapsulate the memory mapping state into its own class):
```cpp
// Prepares for a call to fork() by taking ownership of internal mutexes.
// This is needed since fork() only duplicates the calling thread, which means
// dangling threads that locked the mutex before forking would never unlock it
// in the child process.
```

> I think we need to discuss some way to more consistently and systematically handle both of these cases (signals + fork)

At least as far as the fork() issue goes, it sounds like what we really want to do is systematically wind down all FEX-internal threads before executing the fork(), so the problem doesn't come up at all (which may or may not be architecturally feasible at this point). Or alternatively, we could do all FEX-internal threads to a dedicated external process.